### PR TITLE
Make Migrations Cancelable

### LINKED ISSUE #12845: Migration: add Cancel button
currently you you have to go to `<gitea-url>/<owner>/<migrating-repo>/settings` and use the "Delete Repo" button, to cancel migration

A cancel Button would be nice

### PR DESCRIPTION
**Pull request by 6543**
close #12845

### CODE REVIEW

File: routers/repo/migrate.go
Comment (6543) [inline routers/repo/migrate.go:208]: lunny: "We should send a stop singal to the migration go routine if it's migrating."

Comment (6543) [reply routers/repo/migrate.go:208]: from discord: ..."Basically we'd need a publish/subscribe system (in redis terms)."...

File: templates/repo/migrate/migrating.tmpl
Comment (silverwind) [inline templates/repo/migrate/migrating.tmpl]:
```suggestion
								<img src="{{StaticUrlPrefix}}/img/loading.png"/>
```

File: templates/repo/migrate/migrating.tmpl
Comment (silverwind) [inline templates/repo/migrate/migrating.tmpl]:
```suggestion
					{{if .Permission.IsAdmin}}
						<div class="ui divider"></div>
						<div class="item center">
							<form action="{{.Link}}/migrate_cancel" method="post">
								{{.CsrfTokenHtml}}
								<button class="ui button" data-do="delete">
									<span class="item text">{{.i18n.Tr "cancel"}}</span>
								</button>
							</form>
						</div>
					{{end}}
```

File: routers/repo/migrate.go
Comment (silverwind) [inline routers/repo/migrate.go]:
```suggestion
// CancelMigration cancels a running migration
```

File: routers/routes/routes.go
Comment (silverwind) [inline routers/routes/routes.go]: Maybe `/migrate/cancel`? That style seems a bit more common then underscores.

File: templates/repo/migrate/migrating.tmpl
Comment (6543) [inline templates/repo/migrate/migrating.tmpl]:
```suggestion
							<form action="{{.Link}}/migrate/cancel" method="post">
```

File: routers/routes/routes.go
Comment (6543) [inline routers/routes/routes.go]:
```suggestion
		m.Post("/migrate/cancel", repo.CancelMigration)
```

File: routers/repo/migrate.go
Comment (6543) [reply routers/repo/migrate.go:208]: @lunny since people are going to use "delete repo" anyway after asking in discord & co. - should we add this now and leave a new issue open to refactor the tasks to be able cancel on task level?

File: routers/repo/migrate.go
Comment (6543) [inline routers/repo/migrate.go:208]:
```suggestion

	// TODO: get task from queues and kill it
	// https://github.com/go-gitea/gitea/pull/12917/files#r492109921

```

**Comment by 6543** [comment]
marked as ready, as per https://github.com/go-gitea/gitea/pull/12917/files#r527381336 & 00e532e

**Comment by silverwind** [comment]
What is this `|` in your screenshot?

image

**Comment by 6543** [comment]
> What is this `|` in your screenshot?

the cursur :laughing:

**Comment by silverwind** [comment]
Shouldn't a button show `cursor: pointer`? :)

**Comment by 6543** [comment]
@silverwind if you like to have a "fresh" screenshot (00e532e460e790d5449931ceec0f718435de9261):
Bildschirmfoto zu 2020-11-29 00-58-13

File: routers/repo/migrate.go
Comment (silverwind) [inline routers/repo/migrate.go]: Maybe add a "The migration has been cancelled." here instead.

**Comment by 6543** [comment]
@silverwind done

File: routers/repo/migrate.go
Comment (6543) [reply routers/repo/migrate.go]: @silverwind done

**Review by silverwind — APPROVED** [review]

File: modules/git/repo.go
Comment (6543) [inline modules/git/repo.go]: we should in future should do more gitcomands with ctx

File: modules/git/repo.go
Comment (zeripath) [inline modules/git/repo.go]:
```suggestion
	cmd := NewCommandContextNoGlobals(ctx, args...).AddArguments("clone")
```

File: modules/migrations/gitea_uploader.go
Comment (zeripath) [inline modules/migrations/gitea_uploader.go]: What happened to doer here?

File: routers/repo/migrate.go
Comment (6543) [inline routers/repo/migrate.go]:
```suggestion
	ctx.Flash.Error(ctx.Tr("repo.migrate.task_not_found", ctx.Repo.Repository.FullName()))
```

File: routers/repo/migrate.go
Comment (6543) [inline routers/repo/migrate.go]: we could store taskID in future ... it process is a "task"

File: modules/migrations/gitea_uploader.go
Comment (6543) [reply modules/migrations/gitea_uploader.go]: It's not needed nor used

File: routers/repo/migrate.go
Comment (zeripath) [inline routers/repo/migrate.go]: It'd be better to store the pid on the task table and grab it from there.

Comment (6543) [reply routers/repo/migrate.go]: ... In the end we need UUIDs

File: modules/git/repo.go
Comment (zeripath) [reply modules/git/repo.go]: agreed!

**Comment by 6543** [comment]
@zeripath done (cab5d46)

**Comment by 6543** [comment]
OK, I think **this is ready**

*to less code has moved to be a refactor*

File: models/migrations/migrations.go
Comment (lunny) [inline models/migrations/migrations.go]: It seems these changes are unrelated?

Comment (6543) [reply models/migrations/migrations.go]: they are - should I move it into it's own pull?

File: modules/process/manager.go
Comment (lunny) [inline modules/process/manager.go:67]: The guid will be changed once the Gitea restarted?

Comment (6543) [reply modules/process/manager.go:67]: yes

File: models/migrations/migrations.go
Comment (6543) [reply models/migrations/migrations.go]: moved into  #13789

File: routers/repo/migrate.go
Comment (lunny) [inline routers/repo/migrate.go:235]: So once there is a running task and the gitea instance restarted. The tasks should be set stopped when the instance restarted. But since guid maybe changed, how do I know which running tasks should be stopped?

Comment (6543) [reply routers/repo/migrate.go:235]: since task is stopped, this doesn't mather - if it will be started again after a gitea-restart, it will update those values

File: routers/repo/migrate.go
Comment (lunny) [inline routers/repo/migrate.go:243]: The user will see a migrating task, but when click cancel, gitea will say it's not exist. This is still a problem.

**Comment by 6543** [comment]
Blocked until we have a final decission on how to handle tasks & processes in a cluster

**Comment by 6543** [comment]
Blocked by #13791

Partialy moved to  #13792

File: templates/repo/migrate/migrating.tmpl
Comment (delvh) [inline templates/repo/migrate/migrating.tmpl:32]: Why do you allow only admins to cancel a migration?
Shouldn't you additionally check whether the requesting user is the user that started the migration?
That would make more sense in my opinion.

And if I see this correctly, this check cannot be avoided as it is also possible that outsiders see this template, if they request the repo URL while it is still migrating. It will simply be displayed as migration failed, which in my opinion is an error of its own. I have tested that on `main`.

EDIT: Or do I perhaps misunderstand the `.Permission.IsAdmin` check such that it checks whether the requesting user has admin rights for this repository instead of the whole server? If that is the case, ignore what I said (except for the additional error that is out of scope for this PR).

File: models/migrations/v161.go
Comment (delvh) [inline models/migrations/v161.go:1]: Well, this file certainly needs to be renamed until this PR will be merged.

File: models/task.go
Comment (delvh) [inline models/task.go:126]: Does that compile?
To me, this seems like a syntax error.
This whole function seems not finished.
Could it be that you intended to finish it another time?

Comment (delvh) [reply models/task.go:126]: If it compiles, I can understand why, but I wouldn't have dreamed up such a syntax ever…

File: models/task.go
Comment (delvh) [inline models/task.go:119]: I think simply returning `0, 0` is not always the best idea.
What if you extended the return to produce an additional error (`(guid, pid int64, err error)`)?
Alternatively, you could set the error return value to `-1, -1`.
This is because it is guaranteed that no process will ever have a negative PID, while it could happen that a process has PID 0, although that is close to impossible for a process that is not the OS.

File: routers/repo/migrate.go
Comment (delvh) [inline routers/repo/migrate.go]:
```suggestion
	if !ctx.Repo.IsAdmin() {
```
I think that suffices.

File: routers/repo/migrate.go
Comment (delvh) [inline routers/repo/migrate.go:239]: Is it even sensible to flash a message if the next action is to redirect back to the homepage?

File: routers/repo/migrate.go
Comment (delvh) [reply routers/repo/migrate.go:243]: Is it even sensible to flash a message if the next action is to redirect back to the homepage?

File: routers/repo/migrate.go
Comment (delvh) [inline routers/repo/migrate.go:237]: I think it would be best to delete the repository here explicitly.
Otherwise, the repo might or might not exist in the database,
resulting in the user being unable to recreate a repo under this name until
either an admin runs a health check, or a doctor command is scheduled.

File: templates/repo/migrate/migrating.tmpl
Comment (6543) [reply templates/repo/migrate/migrating.tmpl:32]: yes this check if current user has admin access to that specific repo

File: models/task.go
Comment (6543) [reply models/task.go:126]: it do compile :)

https://play.golang.org/p/Ei7xfQpJFTn

**Comment by singuliere** [comment]
This is an interesting feature, how could the conflicts be resolved? Would you like to rebase or should it be done in a different way?

**Comment by 6543** [comment]
I'll resolve the conflicts ...

**Comment by 6543** [comment]
to explain the problem why it did not got merged jet: if you have a setup with more instances, and you like to cancle a migration from instance A running on instance B, it wont work (A will find the job in the database but cant find it in it's process.Manager)

**Pull request closed without merging**
